### PR TITLE
Add freetype dependencies via tk_variant so that tk builds with freetype fonts

### PR DESCRIPTION
**Pull request by smithsp**
While I understand that the anaconda conda distribution may not be willing to accept these changes to have tk built with freetype support, is there objection from the conda-forge tk package?  If this pull request can not be accepted as it is, I could use some guidance to perhaps create a variant (is this the right word?) that could be opted into.

I just built this on my own system, but I'm not sure if it will build on the conda-forge system - we will see from the auto-checks.  I haven't added anything to the checks to be sure the fonts are included appropriately.

Checklist
* [x] Used a [fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)

### CODE REVIEW

**Comment by mingwandroid** [comment]
I have no objection to this. We might need to revert it for AD though. CF can have a slightly nicer looking TK if the general concensus is that cycles in our dependencies are OK.

Pinging @CJ-Wright, @isuruf on that.

**Comment by isuruf** [comment]
Where's the cycle coming from?

**Comment by mingwandroid** [comment]
AFAIR a Python requirement via one of the font packages, likely font-config.

Now-a-days, it may be possible that python is in the build requirements which would allow us to break the cycle.

**Comment by mingwandroid** [comment]
@smithsp: `run_exports` should add most of these run requirements for you.

**Comment by smithsp** [comment]
@mingwandroid I read the documentation for `run_exports` and it was not clear to me if it could be added to the host section, in addition to the build section.  I see that it now built/tested correctly on linux with the run requirements, so I will make that change and iterate appropriately.

**Comment by mingwandroid** [comment]
Here I am talking about the `run_exports` identified in the packages in `requirements/host`, not the `run_exports` of this package.

The idea is that the packages that are used as your direct dependencies in `requirements/host` themselves carry dependencies and (semver respecting, largely) pinings for those dependencies and these are automatically injected into `requirements/run` by conda-build.

In other words, you shouldn't need most of these, provided the packages identify their `run_exports` correctly (which I believe nearly all do these days).

**Comment by mingwandroid** [comment]
In general, the ideal situation is that recipes should have empty `requirements/run` sections so that `conda-build` can apply them intelligently and consistently.

**Comment by calroc** [comment]
please please please!

I have a project that uses Tkinter and it would be awesome if I could tell people "Just use Conda to get Python", but right now it looks like this:
potato

For better or worse Tkinter is the default shipping GUI system that comes with Python (I know wx ate Tk's lunch but it's still *there*.) I would really appreciate it if it could have fonts that look nice.

**Comment by smithsp** [comment]
@conda-forge-admin, please rerender

**Comment by smithsp** [comment]
@mingwandroid Is the error: `wish: error while loading shared libraries: libXft.so.2: cannot open shared object file: No such file or directory` indicative of the `xorg-libxft` package not configuring its `run_exports` correctly?  Should I add `xorg-libxft` here to the run requirements, or try to open a bugfix on that feedstock?

**Comment by smithsp** [comment]
@conda-forge/tk This PR is now passing, so I would appreciate if further comments could be provided, if any, or this be merged.

**Comment by calroc** [comment]
Please can someone take a look at this?

**Comment by jjhelmus** [comment]
Thanks for seeing this though @smithsp.  I'm in favor of this change but have some reservations about having `tk` require various `xorg-` package.  `tk` is a requirement of `python` these packages will  be required in nearly every user's environment.  

I'm going to get the opinion of @conda-forge/core in today's meeting on this topic.

**Comment by jjhelmus** [comment]
For reference with the change an environment containing only python will require:
```
$ conda create -n cf_test -c local -c conda-forge python
...
The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-0_gnu
  ca-certificates    conda-forge/linux-64::ca-certificates-2020.4.5.1-hecc5488_0
  fontconfig         conda-forge/linux-64::fontconfig-2.13.1-h1056068_1002
  freetype           conda-forge/linux-64::freetype-2.10.2-he06d7ca_0
  icu                conda-forge/linux-64::icu-67.1-he1b5a44_0
  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.34-h53a641e_3
  libffi             conda-forge/linux-64::libffi-3.2.1-he1b5a44_1007
  libgcc-ng          conda-forge/linux-64::libgcc-ng-9.2.0-h24d8f2e_2
  libgomp            conda-forge/linux-64::libgomp-9.2.0-h24d8f2e_2
  libiconv           conda-forge/linux-64::libiconv-1.15-h516909a_1006
  libpng             conda-forge/linux-64::libpng-1.6.37-hed695b0_1
  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-9.2.0-hdf63c60_2
  libuuid            conda-forge/linux-64::libuuid-2.32.1-h14c3975_1000
  libxcb             conda-forge/linux-64::libxcb-1.13-h14c3975_1002
  libxml2            conda-forge/linux-64::libxml2-2.9.10-h72b56ed_1
  ncurses            conda-forge/linux-64::ncurses-6.1-hf484d3e_1002
  openssl            conda-forge/linux-64::openssl-1.1.1g-h516909a_0
  pthread-stubs      conda-forge/linux-64::pthread-stubs-0.4-h14c3975_1001
  python             conda-forge/linux-64::python-3.8.2-he5300dc_7_cpython
  python_abi         conda-forge/linux-64::python_abi-3.8-1_cp38
  readline           conda-forge/linux-64::readline-8.0-hf8c457e_0
  sqlite             conda-forge/linux-64::sqlite-3.30.1-hcee41ef_0
  tk                 home/jhelmus/conda/conda-bld/linux-64::tk-8.6.10-h14c3975_1005
  xorg-kbproto       conda-forge/linux-64::xorg-kbproto-1.0.7-h14c3975_1002
  xorg-libx11        conda-forge/linux-64::xorg-libx11-1.6.9-h516909a_0
  xorg-libxau        conda-forge/linux-64::xorg-libxau-1.0.9-h14c3975_0
  xorg-libxdmcp      conda-forge/linux-64::xorg-libxdmcp-1.1.3-h516909a_0
  xorg-libxft        conda-forge/linux-64::xorg-libxft-2.3.3-h71203ad_0
  xorg-libxrender    conda-forge/linux-64::xorg-libxrender-0.9.10-h516909a_1002
  xorg-renderproto   conda-forge/linux-64::xorg-renderproto-0.11.1-h14c3975_1002
  xorg-xproto        conda-forge/linux-64::xorg-xproto-7.0.31-h14c3975_1007
  xz                 conda-forge/linux-64::xz-5.2.5-h516909a_0
  zlib               conda-forge/linux-64::zlib-1.2.11-h516909a_1006
```

Currently such an environment requires far fewer packages:

```
The following NEW packages will be INSTALLED:

  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-0_gnu
  ca-certificates    conda-forge/linux-64::ca-certificates-2020.4.5.1-hecc5488_0
  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.34-h53a641e_3
  libffi             conda-forge/linux-64::libffi-3.2.1-he1b5a44_1007
  libgcc-ng          conda-forge/linux-64::libgcc-ng-9.2.0-h24d8f2e_2
  libgomp            conda-forge/linux-64::libgomp-9.2.0-h24d8f2e_2
  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-9.2.0-hdf63c60_2
  ncurses            conda-forge/linux-64::ncurses-6.1-hf484d3e_1002
  openssl            conda-forge/linux-64::openssl-1.1.1g-h516909a_0
  python             conda-forge/linux-64::python-3.8.2-he5300dc_7_cpython
  python_abi         conda-forge/linux-64::python_abi-3.8-1_cp38
  readline           conda-forge/linux-64::readline-8.0-hf8c457e_0
  sqlite             conda-forge/linux-64::sqlite-3.30.1-hcee41ef_0
  tk                 conda-forge/linux-64::tk-8.6.10-hed695b0_0
  xz                 conda-forge/linux-64::xz-5.2.5-h516909a_0
  zlib               conda-forge/linux-64::zlib-1.2.11-h516909a_1006
```

**Comment by jjhelmus** [comment]
One alternative is to use a CDT for `libxft` using the following diff:
``` diff
diff --git a/recipe/meta.yaml b/recipe/meta.yaml
index 9e931c4..7c34014 100644
--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -27,14 +27,12 @@ requirements:
     - {{ cdt('libx11-devel') }}          # [linux]
     - {{ cdt('libxcb') }}                # [linux]
     - {{ cdt('libxau') }}                # [linux]
+    - {{ cdt('libxft') }}                # [linux]
     - make                               # [linux]
   host:
     - zlib  # [unix]
     - freetype
     - fontconfig
-    - xorg-libxft  # [not win]
-  run:
-    - xorg-libxft  # [not win]
 
 test:
```

**Comment by smithsp** [comment]
@jjhelmus Thank you for your attention and suggestions.  I will push a commit with the `cdt` change.  (What does `cdt` stand for anyway?)  While it is true that a conda-forge based python environment will require more packages, a base channel bare environment would still be available.  Another proposal I suggested was to make a variant like mpi, but I would need more help there.

**Comment by smithsp** [comment]
@jjhelmus With the cdt change I get
`conda_build.exceptions.DependencyNeedsBuildingError: Unsatisfiable dependencies for platform linux-aarch64: {'libxft-cos7-aarch64'}`
for the `linux_aarch64_` build.  I don't know what to do with that, except revert to what I had before.

**Comment by jjhelmus** [comment]
After looking into this further the `xft` CDT approach will not work, but building against `xorg-libxft` will work.  The key difference between packages which have nice looking fonts and those that do not if Tk it built with [Xft support](https://wiki.tcl-lang.org/page/Xft+support), also discussed on the [Xft page](https://wiki.tcl-lang.org/page/Xft).  

Xft support is enabled automatically if the configuration process is able to find the needed libraries.  In the logs this check is reported in the `checking whether to use xft... ` line.  That needs to be `yes` to get nice looking anti-aliased fonts. Builds which includes `xorg-libxft` as a host requirement xft are able to find the needed libraries, those using the `xft` CDT do not because the CDT only provides the libraries not the headers or associated binaries.

We discussed the additional requirements at this weeks core meeting and there was some concern over adding the xorg and so many additional libraries to every environment containing `python`.  I think a good solution would be to provide two variants of the `tk` package as @smithsp mentioned.  The default would be the same as the existing packages and not have xft support.  A second would include xft support and require the xorg and other libraries.  Users could select between these two variants using a mutex package in a similar manner as they are able choose between blas, python and openmp implementation.  

As for a name for the mutex package I would put forward `tk_variant`, with `_tk_mutex` as an second option.  Initial there would be two builds of this package, `tk_variant=1.0=noxft` and `tk_variant=1.0=xft` with the `xft` having a feature so that `noxft` is the default selection.  Users can select the Xft variant with nicer font by including `tk_variant=*=xft` in their requirement specs.

@conda-forge/core Does this approach and naming scheme seem reasonable?

@smithsp I'm happy to help put this variant based recipe together.  Using multiple outputs it should be possible have all the packages produced by a single recipe.

**Comment by jjhelmus** [comment]
> (What does `cdt` stand for anyway?)

CDT stands for "Core Dependency Tree".  These are a way to reliable use "system libraries" when building conda packages. Some additional details can be found in the [conda build documentation](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html?highlight=CDT#host)

**Comment by beckermr** [comment]
SGTM. One question though.

Do we want to adopt a standard-ish naming scheme for mutex packages? 

This is a common pattern. We have some which are `_<thing>_mutex` and others which are not. I prefer the approach with an underscore personally, but having one across the ecosystem going forward might be good.

**Comment by calroc** [comment]
This would be better than the existing situation, but I think `tk_variant=1.0=xft` should be the default.  Is there a use case for Tk *without* Xft, given what it provides?

> The Xft library provides support for antialiased TrueType fonts, full Unicode, and an un-brain-damaged font selection mechanism for X11.

If you ship TCL/Tk + Tkinter it should work right.  It would be better to make the whole of Tk optional than to ship it broken, I think.

**Comment by smithsp** [comment]
@conda-forge-admin, please rerender

**Comment by bszollosinagy** [comment]
This would be a very useful change, to restore the platform independence of TKinter in Python. It is the GUI included with Python, so making it work in Linux, as intended, seems to be in line with Conda's goals. Codeowners, is your only worry, that now more dependencies will be needed in Linux?  I see that there are my Python builds.. even for the same version... can we simply then have a python build that includes this PR?

**Comment by smithsp** [comment]
Mention @kalling @wdeshazer

**Comment by smithsp** [comment]
@conda-forge/core is going to want to look over this in conjunction with https://github.com/conda-forge/staged-recipes/pull/23868 to introduce a `tk_variant`.
I don't think I understand what @jjhelmus indicated by "Using multiple outputs it should be possible have all the packages produced by a single recipe.", but hopefully I have satisfied it with the current incarnation of the recipe.

**Comment by smithsp** [comment]
Note to self: this is failing until the `tk_variant` package exists.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -15,8 +16,15 @@ source:
...
+
+{% if tk_variant == 'noxft' %}
+# prioritize noxft variant via build number
+{% set build = build + 1000 %}
+{% endif %}
+
 build:
-  number: 0
+  number: {{ build }}
+  string: "{{ tk_variant }}_h{{ PKG_HASH }}_{{ build }}"
```
Comment (isuruf) [inline recipe/meta.yaml:27]:
```suggestion
  string: "{{ tk_variant }}_h{{ PKG_HASH }}_{{ build }}"
  track_features:
    - tk_xft   # [tk_variant == 'xft']
```

**Comment by smithsp** [comment]
@conda-forge-admin, please rerender

**Comment by smithsp** [comment]
@conda-forge/core Is this on your radar for the next core meeting?  Is there something more I need to do for this PR?

**Comment by smithsp** [comment]
@conda-forge-admin, please rerender

**Comment by smithsp** [comment]
@conda-forge-admin, please rerender

**Comment by smithsp** [comment]
@conda-forge/core Has this been discussed internally yet?

**Comment by smithsp** [comment]
@conda-forge/core @conda-forge/tk 
I haven't heard back from any of you since my last ping a month ago.  Any feedback would be appreciated.

**Comment by isuruf** [comment]
Looking at the build logs, I don't see any indication that freetype or fontconfig is being used for compiling/linking.

**Comment by smithsp** [comment]
@isuruf Thank you for being thorough.  You are right that in the log for the xft build
https://dev.azure.com/conda-forge/84710dde-1620-425b-80d0-4cf5baca359d/_apis/build/builds/794110/logs/30
I see the line
`checking whether to use xft... no`
I will go back to see what I need to do differently.

**Comment by smithsp** [comment]
I raised https://github.com/conda-forge/cdt-builds/issues/67 because I don't think the `libxft` CDT provides `libXft` any more.

**Comment by smithsp** [comment]
@jaimergp That is really cool.  The exact file that is being sought is `X11/Xft/Xft.h`, which is not a part of that package.  So I checked `xorg-libxft` (thanks for that suggestion), and it does include that package.  I will change my dependency from `cdt( 'libxft' )` to `xorg-libxft`.

**Comment by smithsp** [comment]
@conda-forge-admin, please rerender

**Comment by jaimergp** [comment]
Looks like you might also need it in `run`.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -30,7 +40,11 @@ requirements:
     - make                               # [linux]
   host:
     - zlib  # [unix]
-
+    - freetype  # [tk_variant == 'xft']
+    - fontconfig  # [tk_variant == 'xft']
+    - xorg-libxft  # [tk_variant == 'xft']
```
Comment (hmaarrfk) [inline recipe/meta.yaml:45]: please consider converting other xorg  CDTs into standard conda packages as well

Comment (isuruf) [reply recipe/meta.yaml:45]: That needs to be a bigger discussion I guess about adding more dependencies to

**Comment by hmaarrfk** [comment]
sorry. I didn't mean to close this. wrong button press.

File: recipe/meta.yaml (hunk 2 above)
Comment (smithsp) [reply recipe/meta.yaml:45]: Actually, I just opened https://github.com/conda-forge/cdt-builds/pull/68 to get a libXft-devel cdt.  Is the team moving away from cdt's, or is there just a concern here for tk, as a dependency of python?

**Comment by smithsp** [comment]
@conda-forge-admin, please rerender

**Comment by smithsp** [comment]
Since https://github.com/conda-forge/cdt-builds/pull/68 did not work out of the box, can we merge this as is, and I can update the recipe when the cdt becomes available?
@isuruf I verified in the build logs that libxft is being used.  Thanks for catching that.

**Comment by smithsp** [comment]
I have converted to use the `libXft-devel` cdt.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -13,8 +14,17 @@ source:
...
+
+{% if tk_variant == 'noxft' %}
+# prioritize noxft variant via build number
+{% set build = build + 1000 %}
+{% endif %}
+
 build:
-  number: 0
+  number: {{ build }}
+  string: "{{ tk_variant }}_h{{ PKG_HASH }}_{{ build }}"
```
Comment (isuruf) [inline recipe/meta.yaml:25]:
```suggestion
  string: "{{ tk_variant }}_h{{ PKG_HASH }}_{{ build }}"     # [;linux]
```

**Review by isuruf — APPROVED** [review]

File: recipe/meta.yaml [hunk 4]
```diff
@@ -1,5 +1,6 @@
 {% set version = "8.6.13" %}
 {% set maj_min = ".".join(version.split(".")[:2]) %}
+{% set build = 1005 %}
```
Comment (isuruf) [inline recipe/meta.yaml:3]:
```suggestion
{% set build = 1 %}
```

File: recipe/meta.yaml [hunk 5]
```diff
@@ -13,8 +14,17 @@ source:
     folder: tk{{ version }}
     sha256: 2e65fa069a23365440a3c56c556b8673b5e32a283800d8d9b257e3f584ce0675
 
+
+{% if tk_variant == 'noxft' %}
+# prioritize noxft variant via build number
+{% set build = build + 1000 %}
```
Comment (isuruf) [inline recipe/meta.yaml:20]:
```suggestion
{% set build = build + 100 %}
```

File: recipe/meta.yaml [hunk 6]
```diff
@@ -13,8 +14,17 @@ source:
     folder: tk{{ version }}
     sha256: 2e65fa069a23365440a3c56c556b8673b5e32a283800d8d9b257e3f584ce0675
 
+
+{% if tk_variant == 'noxft' %}
```
Comment (isuruf) [inline recipe/meta.yaml:18]:
```suggestion
{% if tk_variant == 'noxft' and linux %}
```

File: recipe/meta.yaml [hunk 7]
```diff
@@ -13,8 +14,17 @@ source:
...
+# prioritize noxft variant via build number
+{% set build = build + 100 %}
+{% endif %}
+
 build:
-  number: 0
+  number: {{ build }}
+  string: "{{ tk_variant }}_h{{ PKG_HASH }}_{{ build }}"     # [linux]
+  track_features:
+    - tk_xft   # [tk_variant == 'xft']
```
Comment (isuruf) [inline recipe/meta.yaml:27]:
```suggestion
    - tk_xft   # [tk_variant == 'xft' and linux]
```

**Comment by smithsp** [comment]
@conda-forge-admin, please rerender

**Comment by isuruf** [comment]
We have `checking whether to use xft... no` on the xft build.

**Comment by smithsp** [comment]
Hmm...  I will investigate.

**Comment by smithsp** [comment]
I'm really confused.  In https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=817131&view=logs&j=d865bc6a-7310-5498-b3ce-1fd883ae7bda&t=9f70c0c7-a4bd-59a9-2151-c0df6a7252b5&l=1973 it says `checking whether to use xft... no`.  In the configure script (line 10089), this should only appear if `tk_aqua` is `no`, but in the `build.sh` we set `--enable-aqua=yes`, which sets `tk_aqua` as yes.   The logs show nothing about changing anything with regard to aqua.  Furthermore if it gets to printing, then the default should be set to `default` not no.  Any thoughts @isuruf ?
Also, when I run the build script on my system, the enable_xft is yes, not no, and finds the `Xft.h` that was missing before.

**Comment by isuruf** [comment]
@conda-forge-admin, rerender

**Comment by smithsp** [comment]
@isuruf Thank you so much for your help.

**Comment by bszollosinagy** [comment]
Thank you all for your work! Do you know how we can activate this on the user side? Is it as simple as "conda install tkinter -c conda-forge" or something similar?

**Comment by azazellochg** [comment]
> Thank you all for your work! Do you know how we can activate this on the user side? Is it as simple as "conda install tkinter -c conda-forge" or something similar?

```conda install -c conda-forge tk=8.6.13=xft_h2e05e87_1``` works for me

**Comment by isuruf** [comment]
`conda install -c conda-forge tk=*=xft_*` works too.

**Comment by bszollosinagy** [comment]
@smithsp Thank you very much, it works like a charm! Even inside a Docker/Conda sandwitch!

**Comment by smithsp** [comment]
@bszollosinagy You're welcome; however, as usual with these things, I got the ball rolling, but as a novice, I got stuck and @isuruf as the pro finished it off, so thanks to @isuruf as well.

**Comment by Arcitec** [comment]
> `conda install -c conda-forge tk=*=xft_*` works too.

## Here's an improved command.

The issue is that Conda installs python and a non-Xft capable `tk` package by default.

We must tell it to create an environment that uses conda-forge and to explicitly request the build of tk that has Xft support (it is not the default). There's also tons of other packages that a default `conda create` installs which we should use conda-forge variants of instead, to ensure full compatibility.

Therefore, we should create a CLEAN Conda environment from conda-forge directly, rather than just adding `tk` to a normal Conda-defaults env later.

To do a **fresh** install with **all** components perfectly installed from conda-forge immediately, we need **this** command:

`conda create -y --prefix "conda_env" -c conda-forge "python==3.10.*" "tk[build=xft_*]"`

This creates an environment in `./conda_env` and installs all packages, along with the Xft-capable version of tk. Xft is what allows Tk to detect your system fonts, render with antialiasing, Unicode support, etc. :)
